packages/plugin-hardhat/src/verify-proxy.ts

```diff
@@ -343,7 +384,71 @@ async function searchEvent(
 }
 
 /**
- * Verifies a contract by looking up an event that should have been logged during contract construction,
+ * Verifies a contract by matching with known artifacts.
+ *
+ * If a match was not found, falls back to verify directly using the regular hardhat verify task.
+ *
+ * If the fallback passes, logs as success.
+ * If the fallback also fails, records errors for both the original and fallback attempts.
+ *
+ * @param hre
+ * @param etherscanApi The Etherscan API config
+ * @param address The contract address to verify
+ * @param possibleContractInfo An array of possible contract artifacts to use for verification along
+ *  with the corresponding creation event expected in the logs.
+ * @param errorReport Accumulated verification errors
+ * @param convertErrorsToWarningsOnFallbackSuccess If fallback verification occurred and succeeded, whether any
+ *  previously accumulated errors should be converted into warnings in the final summary.
+ * @throws {EventNotFound} if none of the events were found in the contract's logs according to Etherscan.
+ */
+async function verifyWithArtifactOrFallback(
+  hre: HardhatRuntimeEnvironment,
+  hardhatVerify: (address: string) => Promise<any>,
+  etherscanApi: EtherscanAPIConfig,
+  address: string,
+  possibleContractInfo: VerifiableContractInfo[],
+  errorReport: ErrorReport,
+  convertErrorsToWarningsOnFallbackSuccess: boolean,
+) {
+  try {
+    await attemptVerifyWithCreationEvent(hre, etherscanApi, address, possibleContractInfo, errorReport);
+    return true;
+  } catch (origError: any) {
+    if (origError instanceof BytecodeNotMatchArtifact || origError instanceof EventNotFound) {
+      // Try falling back to regular hardhat verify in case the source code is available in the user's project.
+      try {
+        await hardhatVerify(address);
+        console.log(`Successfully verified contract at ${address}.`);
+      } catch (fallbackError: any) {
+        if (fallbackError.message.toLowerCase().includes('already verified')) {
+          console.log(`Contract at ${address} already verified.`);
```

Comment:
Again I would expect some sort of message of this kind to already be printed by hardhat-verify itself. Doesn't it do that?

Reply:
It throws an error with 'Contract source code already verified', but we want to catch it (not fail) and log with the address for context.